{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data Preprocessing\n",
    "\n",
    "The data preprocessing stage consists of the following steps:\n",
    "\n",
    "1. creating numerical encodings for the influenza protein sequences,\n",
    "2. splitting the data into training, validation, and testing.\n",
    "\n",
    "We consider two encoding schemes: one-hot and $k$-mer. For each one, a different dataset is generated.\n",
    "\n",
    "The partitions into training, validation, and testing are done beforehand so that no algorithm benefits from potential special \"tweaking\" due to partitioning the datasets differently."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "# SADR: import all auxiliary functions.\n",
    "from utils import *\n",
    "\n",
    "from sklearn.utils import shuffle\n",
    "import pickle"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Loading Data\n",
    "\n",
    "We begin by parsing the `*.fasta` files corresponding to the sequences into `DataFrames`. The most relevant metadata is extracted. The labels `-1` for the animal and `+1` for the human sequences are also assigned."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "# SADR: path to data sets.\n",
    "animal_sequences = os.path.join(\"data_animal\", \"sequences_20250325_2832148.fasta\")\n",
    "human_sequences = os.path.join(\"data_human\", \"sequences_20250325_9135018.fasta\")\n",
    "\n",
    "# SADR: reading datasets to memory.\n",
    "df_animal, seqs_animal = parse_fasta(animal_sequences, -1)\n",
    "df_human, seqs_human = parse_fasta(human_sequences, 1)\n",
    "\n",
    "# SADR: creating a single dataframe with all sequences.\n",
    "df_sequences = pd.concat([df_animal, df_human], ignore_index=True)\n",
    "df_sequences[\"sequence\"] = seqs_animal + seqs_human"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_id</th>\n",
       "      <th>virus_info</th>\n",
       "      <th>location</th>\n",
       "      <th>host</th>\n",
       "      <th>label</th>\n",
       "      <th>is_partial</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABV48048.1</td>\n",
       "      <td>hemagglutinin, partial [Influenza A virus (A/c...</td>\n",
       "      <td>China</td>\n",
       "      <td>Alectoris chukar</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>DKICIGYQSTNSTETVDTLTETNVPVTQAKELLHTEHNGMLCATNL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>QZF77349.1</td>\n",
       "      <td>hemagglutinin [Influenza A virus]</td>\n",
       "      <td>Uganda</td>\n",
       "      <td>Gallus gallus</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>MEAIPLLTLLLVVTTSSADKICVGHQSTNSTETVDTLTEANVPVTQ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AWV94998.1</td>\n",
       "      <td>hemagglutinin [Influenza A virus]</td>\n",
       "      <td>USA</td>\n",
       "      <td>Arenaria interpres</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>MEAKLFVLFCTFTVLKADTICVGYHANNSTDTVDTVLEKNVTVTHS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACM17970.1</td>\n",
       "      <td>hemagglutinin [Influenza A virus (A/chicken/Ne...</td>\n",
       "      <td>USA</td>\n",
       "      <td>Gallus gallus</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>MNIQILAFIACVLTGAKGDKICLGHHAVANGTKVNTLTEKGIEVVN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACR27188.1</td>\n",
       "      <td>hemagglutinin, partial [Influenza A virus (A/t...</td>\n",
       "      <td>France</td>\n",
       "      <td>Meleagris gallopavo</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>ETAYVSVVASHYNRRFTPEIARRPKIRDQEGRINYYWTLLEPSDTI...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       seq_id                                         virus_info location  \\\n",
       "0  ABV48048.1  hemagglutinin, partial [Influenza A virus (A/c...    China   \n",
       "1  QZF77349.1                  hemagglutinin [Influenza A virus]   Uganda   \n",
       "2  AWV94998.1                  hemagglutinin [Influenza A virus]      USA   \n",
       "3  ACM17970.1  hemagglutinin [Influenza A virus (A/chicken/Ne...      USA   \n",
       "4  ACR27188.1  hemagglutinin, partial [Influenza A virus (A/t...   France   \n",
       "\n",
       "                  host  label  is_partial  \\\n",
       "0     Alectoris chukar     -1           1   \n",
       "1        Gallus gallus     -1           0   \n",
       "2   Arenaria interpres     -1           0   \n",
       "3        Gallus gallus     -1           0   \n",
       "4  Meleagris gallopavo     -1           1   \n",
       "\n",
       "                                            sequence  \n",
       "0  DKICIGYQSTNSTETVDTLTETNVPVTQAKELLHTEHNGMLCATNL...  \n",
       "1  MEAIPLLTLLLVVTTSSADKICVGHQSTNSTETVDTLTEANVPVTQ...  \n",
       "2  MEAKLFVLFCTFTVLKADTICVGYHANNSTDTVDTVLEKNVTVTHS...  \n",
       "3  MNIQILAFIACVLTGAKGDKICLGHHAVANGTKVNTLTEKGIEVVN...  \n",
       "4  ETAYVSVVASHYNRRFTPEIARRPKIRDQEGRINYYWTLLEPSDTI...  "
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_sequences.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_id</th>\n",
       "      <th>virus_info</th>\n",
       "      <th>location</th>\n",
       "      <th>host</th>\n",
       "      <th>label</th>\n",
       "      <th>is_partial</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3995</th>\n",
       "      <td>UWV21073.1</td>\n",
       "      <td>hemagglutinin [Influenza A virus]</td>\n",
       "      <td>USA</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>MKTIIALSNILCLVFAQKIPGNDNSTATLCLGHHAVPNGTIVKTIT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3996</th>\n",
       "      <td>UUV82183.1</td>\n",
       "      <td>hemagglutinin [Influenza A virus]</td>\n",
       "      <td>USA</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>MKTIIALSNILCLVFAQKIPGNDNSTATLCLGHHAVPNGTVVKTIT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3997</th>\n",
       "      <td>WOP34589.1</td>\n",
       "      <td>hemagglutinin [Influenza A virus]</td>\n",
       "      <td>USA</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>MKAILVVMLYTFTTANADTLCIGYHANNSTDTVDTVLEKNVTVTHS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3998</th>\n",
       "      <td>WBR18717.1</td>\n",
       "      <td>hemagglutinin [Influenza A virus]</td>\n",
       "      <td>Germany</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>MKTIIALSNILCLVFAQKIPGNDNSTATLCLGHHAVPNGTIVKTIT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3999</th>\n",
       "      <td>ABI21112.1</td>\n",
       "      <td>hemagglutinin [Influenza A virus (A/Wellington...</td>\n",
       "      <td>New Zealand</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>MKTIIALSYILCLVFAQKLPGNDNSTATLCLGHHAVPNGTIVKTIT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          seq_id                                         virus_info  \\\n",
       "3995  UWV21073.1                  hemagglutinin [Influenza A virus]   \n",
       "3996  UUV82183.1                  hemagglutinin [Influenza A virus]   \n",
       "3997  WOP34589.1                  hemagglutinin [Influenza A virus]   \n",
       "3998  WBR18717.1                  hemagglutinin [Influenza A virus]   \n",
       "3999  ABI21112.1  hemagglutinin [Influenza A virus (A/Wellington...   \n",
       "\n",
       "         location          host  label  is_partial  \\\n",
       "3995          USA  Homo sapiens      1           0   \n",
       "3996          USA  Homo sapiens      1           0   \n",
       "3997          USA  Homo sapiens      1           0   \n",
       "3998      Germany  Homo sapiens      1           0   \n",
       "3999  New Zealand  Homo sapiens      1           0   \n",
       "\n",
       "                                               sequence  \n",
       "3995  MKTIIALSNILCLVFAQKIPGNDNSTATLCLGHHAVPNGTIVKTIT...  \n",
       "3996  MKTIIALSNILCLVFAQKIPGNDNSTATLCLGHHAVPNGTVVKTIT...  \n",
       "3997  MKAILVVMLYTFTTANADTLCIGYHANNSTDTVDTVLEKNVTVTHS...  \n",
       "3998  MKTIIALSNILCLVFAQKIPGNDNSTATLCLGHHAVPNGTIVKTIT...  \n",
       "3999  MKTIIALSYILCLVFAQKLPGNDNSTATLCLGHHAVPNGTIVKTIT...  "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_sequences.tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Encoding\n",
    "\n",
    "### One-hot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_id</th>\n",
       "      <th>virus_info</th>\n",
       "      <th>location</th>\n",
       "      <th>host</th>\n",
       "      <th>label</th>\n",
       "      <th>is_partial</th>\n",
       "      <th>sequence</th>\n",
       "      <th>one_hot</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABV48048.1</td>\n",
       "      <td>hemagglutinin, partial [Influenza A virus (A/c...</td>\n",
       "      <td>China</td>\n",
       "      <td>Alectoris chukar</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>DKICIGYQSTNSTETVDTLTETNVPVTQAKELLHTEHNGMLCATNL...</td>\n",
       "      <td>[[0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>QZF77349.1</td>\n",
       "      <td>hemagglutinin [Influenza A virus]</td>\n",
       "      <td>Uganda</td>\n",
       "      <td>Gallus gallus</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>MEAIPLLTLLLVVTTSSADKICVGHQSTNSTETVDTLTEANVPVTQ...</td>\n",
       "      <td>[[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AWV94998.1</td>\n",
       "      <td>hemagglutinin [Influenza A virus]</td>\n",
       "      <td>USA</td>\n",
       "      <td>Arenaria interpres</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>MEAKLFVLFCTFTVLKADTICVGYHANNSTDTVDTVLEKNVTVTHS...</td>\n",
       "      <td>[[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACM17970.1</td>\n",
       "      <td>hemagglutinin [Influenza A virus (A/chicken/Ne...</td>\n",
       "      <td>USA</td>\n",
       "      <td>Gallus gallus</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>MNIQILAFIACVLTGAKGDKICLGHHAVANGTKVNTLTEKGIEVVN...</td>\n",
       "      <td>[[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACR27188.1</td>\n",
       "      <td>hemagglutinin, partial [Influenza A virus (A/t...</td>\n",
       "      <td>France</td>\n",
       "      <td>Meleagris gallopavo</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>ETAYVSVVASHYNRRFTPEIARRPKIRDQEGRINYYWTLLEPSDTI...</td>\n",
       "      <td>[[0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       seq_id                                         virus_info location  \\\n",
       "0  ABV48048.1  hemagglutinin, partial [Influenza A virus (A/c...    China   \n",
       "1  QZF77349.1                  hemagglutinin [Influenza A virus]   Uganda   \n",
       "2  AWV94998.1                  hemagglutinin [Influenza A virus]      USA   \n",
       "3  ACM17970.1  hemagglutinin [Influenza A virus (A/chicken/Ne...      USA   \n",
       "4  ACR27188.1  hemagglutinin, partial [Influenza A virus (A/t...   France   \n",
       "\n",
       "                  host  label  is_partial  \\\n",
       "0     Alectoris chukar     -1           1   \n",
       "1        Gallus gallus     -1           0   \n",
       "2   Arenaria interpres     -1           0   \n",
       "3        Gallus gallus     -1           0   \n",
       "4  Meleagris gallopavo     -1           1   \n",
       "\n",
       "                                            sequence  \\\n",
       "0  DKICIGYQSTNSTETVDTLTETNVPVTQAKELLHTEHNGMLCATNL...   \n",
       "1  MEAIPLLTLLLVVTTSSADKICVGHQSTNSTETVDTLTEANVPVTQ...   \n",
       "2  MEAKLFVLFCTFTVLKADTICVGYHANNSTDTVDTVLEKNVTVTHS...   \n",
       "3  MNIQILAFIACVLTGAKGDKICLGHHAVANGTKVNTLTEKGIEVVN...   \n",
       "4  ETAYVSVVASHYNRRFTPEIARRPKIRDQEGRINYYWTLLEPSDTI...   \n",
       "\n",
       "                                             one_hot  \n",
       "0  [[0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,...  \n",
       "1  [[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,...  \n",
       "2  [[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,...  \n",
       "3  [[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,...  \n",
       "4  [[0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0,...  "
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_sequences[\"one_hot\"] = df_sequences[\"sequence\"].apply(lambda x: one_hot_encoding(x))\n",
    "df_sequences.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "# SADR: creating a vector for each matrix representing the one-hot encoding of the sequences.\n",
    "# The vectors for all training instances are stored in a single matrix.\n",
    "X = np.array([seq.flatten() for seq in df_sequences[\"one_hot\"]])\n",
    "\n",
    "# SADR: vector with labels.\n",
    "y = df_sequences[\"label\"].values\n",
    "\n",
    "# SADR: shuffling the data.\n",
    "X_shuffled, y_shuffled = shuffle(X, y, random_state=42)\n",
    "\n",
    "# SADR: splitting into training and testing (80/20 proportion).\n",
    "split_idx = int(0.8 * len(X_shuffled))\n",
    "X_train_full, y_train_full = X_shuffled[:split_idx], y_shuffled[:split_idx]\n",
    "X_test, y_test = X_shuffled[split_idx:], y_shuffled[split_idx:]\n",
    "\n",
    "# SADR: splitting the training set into training and validation (80/20 proportion).\n",
    "split_idx = int(0.8 * len(X_train_full))\n",
    "X_train, y_train = X_train_full[:split_idx], y_train_full[:split_idx]\n",
    "X_val, y_val = X_train_full[split_idx:], y_train_full[split_idx:]\n",
    "\n",
    "# SADR: saving the training data in disk.\n",
    "dataset_one_hot = {\"X_train\": X_train, \"y_train\": y_train, \"X_val\": X_val, \"y_val\": y_val, \"X_test\": X_test, \"y_test\": y_test}\n",
    "_save_path = os.path.join(\"preprocessed_datasets\", \"dataset_one_hot.pkl\")\n",
    "with open(_save_path, 'wb') as f:\n",
    "    pickle.dump(dataset_one_hot, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### $k$-mer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "# SADR: getting the k-mer encoding of the sequences.\n",
    "encoding, dict = k_mer_encoding(df_sequences[\"sequence\"].values, 2)\n",
    "\n",
    "# SADR: separating features and labels.\n",
    "X = encoding\n",
    "y = df_sequences[\"label\"].values\n",
    "\n",
    "# SADR: shuffling the data.\n",
    "X_shuffled, y_shuffled = shuffle(X, y, random_state=42)\n",
    "\n",
    "# SADR: splitting into training and testing (80/20 proportion).\n",
    "split_idx = int(0.8 * len(X_shuffled))\n",
    "X_train_full, y_train_full = X_shuffled[:split_idx], y_shuffled[:split_idx]\n",
    "X_test, y_test = X_shuffled[split_idx:], y_shuffled[split_idx:]\n",
    "\n",
    "# SADR: splitting the training set into training and validation (80/20 proportion).\n",
    "split_idx = int(0.8 * len(X_train_full))\n",
    "X_train, y_train = X_train_full[:split_idx], y_train_full[:split_idx]\n",
    "X_val, y_val = X_train_full[split_idx:], y_train_full[split_idx:]\n",
    "\n",
    "# SADR: saving the training data in disk.\n",
    "dataset_k_mer = {\"X_train\": X_train, \"y_train\": y_train, \"X_val\": X_val, \"y_val\": y_val, \"X_test\": X_test, \"y_test\": y_test}\n",
    "_save_path = os.path.join(\"preprocessed_datasets\", \"dataset_k_mer.pkl\")\n",
    "with open(_save_path, 'wb') as f:\n",
    "    pickle.dump(dataset_k_mer, f)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
